SignalServiceKit/src/Messages/Interactions/TSQuotedMessage.h

```diff
@@ -15,25 +51,49 @@ NS_ASSUME_NONNULL_BEGIN
 // or attachment with caption.
 @property (nullable, nonatomic, readonly) NSString *body;
 
-// This property should be set IFF we are quoting an attachment message.
-@property (nullable, nonatomic, readonly) NSString *sourceFilename;
-// This property can be set IFF we are quoting an attachment message, but it is optional.
-@property (nullable, nonatomic, readonly) NSData *thumbnailData;
+#pragma mark - Attachments
+
 // This is a MIME type.
 //
 // This property should be set IFF we are quoting an attachment message.
-@property (nullable, nonatomic, readonly) NSString *contentType;
+- (nullable NSString *)contentType;
+- (nullable NSString *)sourceFilename;
+
+// References a yet-to-be downloaded thumbnail file
+- (nullable NSString *)thumbnailAttachmentPointerId;
+
+// References an already downloaded or locally generated thumbnail file
+- (nullable NSString *)thumbnailAttachmentStreamId;
+- (void)setThumbnailAttachmentStream:(TSAttachment *)thumbnailAttachmentStream;
+
+// currently only used by orphan attachment cleaner
+- (NSArray<NSString *> *)thumbnailAttachmentStreamIds;
+
+@property (atomic, readonly) NSArray<OWSAttachmentInfo *> *quotedAttachments;
+
+// Before sending, persist a thumbnail attachment derived from the quoted attachment
+- (NSArray<TSAttachmentStream *> *)createThumbnailAttachmentsIfNecessaryWithTransaction:
+    (YapDatabaseReadWriteTransaction *)transaction;
 
 - (instancetype)init NS_UNAVAILABLE;
 
+// used when receiving quoted messages
+- (instancetype)initWithTimestamp:(uint64_t)timestamp
+                         authorId:(NSString *)authorId
+                             body:(NSString *_Nullable)body
+    receivedQuotedAttachmentInfos:(NSArray<OWSAttachmentInfo *> *)attachmentInfos;
+
+// used when sending quoted messages
 - (instancetype)initWithTimestamp:(uint64_t)timestamp
                          authorId:(NSString *)authorId
                              body:(NSString *_Nullable)body
-                   sourceFilename:(NSString *_Nullable)sourceFilename
-                    thumbnailData:(NSData *_Nullable)thumbnailData
-                      contentType:(NSString *_Nullable)contentType;
+      quotedAttachmentsForSending:(NSArray<TSAttachment *> *)attachments;
```

Comment:
When would there be multiple attachments?

Reply:
I have no idea. 

The protos support multiple attachments and Moxie keeps threatening for this to be a reality, so I tried to accommodate it on the receiving side while adding accessors that just deal with the "first" instance since that will be the normal case.